Gym/Blackjack/main.swift

```diff
@@ -0,0 +1,180 @@
+// Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Python
+import TensorFlow
+
+let gym = Python.import("gym")
+let environment = gym.make("Blackjack-v0")
+
+let totalIterations = 10000
+let learningPhase = totalIterations * 5 / 100
+
+class BlackjackState {
+    var playerSum: Int = 0
+    var dealerCard: Int = 0
+    var useableAce: Int = 0
+
+    init(pythonState: PythonObject) {
+        self.playerSum = Int(pythonState[0]) ?? 0
+        self.dealerCard = Int(pythonState[1]) ?? 0
+        self.useableAce = Int(pythonState[2]) ?? 0
+    }
+}
+
+enum SolverType: CaseIterable {
+    case random, markov, qlearning, normal
+}
+
+class Solver {
+    var Q: [[[[Float]]]] = []
+    var alpha: Float = 0.5
+    let gamma: Float = 0.2
+
+    let numberOfPlayerStates = 32 // 21 + 10 + 1 offset
+    let numberOfDealerVisibleStates = 11 // 10 + 1 offset
+    let numberOfAceStates = 2 // useable / not bool
+    let numberOfPlayerActions = 2 // hit / stay
+
+    init() {
+        Q = Array(repeating: Array(repeating: Array(repeating: Array(repeating: 0.0,
+                                                                     count: numberOfPlayerActions),
+                                                    count: numberOfAceStates),
+                                   count: numberOfDealerVisibleStates),
+                  count: numberOfPlayerStates)
+    }
+
+    func updateQLearningStrategy(prior: BlackjackState,
+                                 action: Int,
+                                 reward: Int,
+                                 post: BlackjackState) {
+        let oldQ = Q[prior.playerSum][prior.dealerCard][prior.useableAce][action]
+        let priorQ = (1 - alpha) * oldQ
+
+        let maxReward = max(Q[post.playerSum][post.dealerCard][post.useableAce][0],
+                            Q[post.playerSum][post.dealerCard][post.useableAce][1])
+        let postQ = alpha * (Float(reward) + gamma * maxReward)
+
+        Q[prior.playerSum][prior.dealerCard][prior.useableAce][action] += priorQ + postQ
+    }
+
+    func getQLearningStrategy(observation: BlackjackState, iteration: Int) -> Bool {
+        let hitReward = Q[observation.playerSum][observation.dealerCard][observation.useableAce][0]
+        let stayReward = Q[observation.playerSum][observation.dealerCard][observation.useableAce][1]
+
+        if (iteration < Int.random(in: 1...learningPhase)) {
```

Comment:
```suggestion
        if iteration < Int.random(in: 1...learningPhase) {
```
Parentheses around conditions are redundant.

Reply:
removed!